certificateutil/filter.go

```diff
@@ -13,22 +13,40 @@ func FilterCertificateInfoModelsByFilterFunc(certificates []CertificateInfoModel
 	return filteredCertificates
 }
 
-// FilterValidCertificateInfos ...
-func FilterValidCertificateInfos(certificateInfos []CertificateInfoModel) []CertificateInfoModel {
+// ValidCertificateInfo contains the certificate infos filtered as valid, invalid and duplicated common name certificates
+type ValidCertificateInfo struct {
+	ValidCertificates,
+	InvalidCertificates,
+	DuplicatedCertificates []CertificateInfoModel
+}
+
+// FilterValidCertificateInfos filters out invalid and duplicated common name certificaates
+func FilterValidCertificateInfos(certificateInfos []CertificateInfoModel) ValidCertificateInfo {
 	certificateInfosByName := map[string]CertificateInfoModel{}
 
+	var invalidCertificates, duplicatedCertificates []CertificateInfoModel
 	for _, certificateInfo := range certificateInfos {
-		if certificateInfo.CheckValidity() == nil {
-			activeCertificate, ok := certificateInfosByName[certificateInfo.CommonName]
-			if !ok || certificateInfo.EndDate.After(activeCertificate.EndDate) {
-				certificateInfosByName[certificateInfo.CommonName] = certificateInfo
-			}
+		if certificateInfo.CheckValidity() != nil {
+			invalidCertificates = append(invalidCertificates, certificateInfo)
+			continue
+		}
+		activeCertificate, ok := certificateInfosByName[certificateInfo.CommonName]
+		if !ok {
+			certificateInfosByName[certificateInfo.CommonName] = certificateInfo
+		} else if certificateInfo.EndDate.After(activeCertificate.EndDate) {
```

Comment:
what if a cert with common name A expires 2016.01.02, found first, then
cert with common name A expires 2016.01.01 <- this will not be added to the `duplicatedCertificates`, since the expire is before the first cert's expire

Reply:
fixed